File: smoke_tests/kotlin/null_safety.kt
// Null Safety and Exception vulnerabilities in Kotlin
package com.example.security

class NullSafetyVulnerabilities {

    // Test 1: Force unwrap with !!
    fun getUserName(userId: Int): String {
        val user = findUser(userId)
        // VULNERABLE: Force unwrap can throw NPE
        return user!!.name
    }

    // Test 2: Unsafe cast
    fun processResponse(data: Any): Map<String, Any> {
        // VULNERABLE: Unsafe cast can throw ClassCastException
        return data as Map<String, Any>
    }

    // Test 3: lateinit without initialization check
    lateinit var database: Database

    fun query(sql: String): List<Any> {
        // VULNERABLE: May throw UninitializedPropertyAccessException
        return database.execute(sql)
    }

    // Test 4: Array index without bounds check
    fun getItem(items: Array<String>, index: Int): String {
        // VULNERABLE: No bounds check
        return items[index]
    }

    // Test 5: First element without check
    fun getFirstUser(users: List<User>): User {
        // VULNERABLE: List might be empty
        return users.first()
    }

    // Test 6: Map get with !!
    fun getConfig(key: String): String {
        val config = loadConfig()
        // VULNERABLE: Key may not exist
        return config[key]!!
    }

    // Test 7: Platform types from Java
    fun processJavaObject(obj: JavaObject): String {
        // VULNERABLE: Platform type could be null
        return obj.getName().uppercase()
    }

    // Test 8: Double bang chain
    fun getNestedValue(data: Map<String, Any?>): String {
        // VULNERABLE: Multiple force unwraps
        return ((data["user"] as Map<String, Any?>)["profile"] as Map<String, Any?>)["name"]!!.toString()
    }

    // Test 9: let with side effect
    fun processOptional(value: String?) {
        // VULNERABLE: value could change between check and use in multi-threaded context
        if (value != null) {
            Thread.sleep(1)
            println(value.length) // value could be null if var
        }
    }

    // Test 10: Regex without match check
    fun extractNumber(input: String): String {
        val regex = Regex("\\d+")
        // VULNERABLE: find() might return null
        return regex.find(input)!!.value
    }

    // Test 11: Iterator next without hasNext
    fun getNext(items: Iterator<String>): String {
        // VULNERABLE: No hasNext check
        return items.next()
    }

    // Test 12: Single element from empty collection
    fun getSingle(items: List<String>): String {
        // VULNERABLE: Throws if not exactly one element
        return items.single()
    }

    // Test 13: Substring without bounds check
    fun extractPart(str: String, start: Int, end: Int): String {
        // VULNERABLE: IndexOutOfBoundsException
        return str.substring(start, end)
    }

    // Test 14: Division without zero check
    fun divide(a: Int, b: Int): Int {
        // VULNERABLE: Division by zero
        return a / b
    }

    // Test 15: Generic type erasure
    inline fun <reified T> parseJson(json: String): T {
        // VULNERABLE: Type mismatch at runtime
        return com.google.gson.Gson().fromJson(json, T::class.java)
    }

    private fun findUser(userId: Int): User? = null
    private fun loadConfig(): Map<String, String> = emptyMap()
}

data class User(val name: String)
class Database {
    fun execute(sql: String): List<Any> = emptyList()
}
interface JavaObject {
    fun getName(): String?
}

File: smoke_tests/scala/null_safety.scala
// Null Safety vulnerabilities in Scala
package com.example.security

class NullSafetyVulnerabilities {

  // Test 1: Option.get without check
  def getUserName(userId: Int): String = {
    val user = findUser(userId)
    // VULNERABLE: get on Option can throw
    user.get.name
  }

  // Test 2: Unsafe pattern matching
  def processOption(opt: Option[String]): String = opt match {
    // VULNERABLE: Missing None case
    case Some(value) => value
    // No None case - will throw MatchError
  }

  // Test 3: Java interop null
  def processJavaObject(obj: JavaObject): String = {
    // VULNERABLE: Java method might return null
    obj.getName.toUpperCase
  }

  // Test 4: head on empty collection
  def getFirstUser(users: List[User]): User = {
    // VULNERABLE: head throws on empty list
    users.head
  }

  // Test 5: last on empty collection
  def getLastItem[T](items: Seq[T]): T = {
    // VULNERABLE: last throws on empty seq
    items.last
  }

  // Test 6: Array index without bounds check
  def getItem(items: Array[String], index: Int): String = {
    // VULNERABLE: No bounds check
    items(index)
  }

  // Test 7: Map apply without contains check
  def getConfig(key: String): String = {
    val config = loadConfig()
    // VULNERABLE: apply throws if key missing
    config(key)
  }

  // Test 8: reduce on empty collection
  def sumItems(items: List[Int]): Int = {
    // VULNERABLE: reduce throws on empty
    items.reduce(_ + _)
  }

  // Test 9: Unsafe cast
  def processResponse(data: Any): Map[String, Any] = {
    // VULNERABLE: Cast can fail
    data.asInstanceOf[Map[String, Any]]
  }

  // Test 10: Pattern match with null
  def matchNull(value: String): String = value match {
    case "test" => "found"
    // VULNERABLE: null not handled
    case other => other.toUpperCase
  }

  // Test 11: orNull usage
  def getNullable(opt: Option[String]): String = {
    // VULNERABLE: Returning null in Scala
    opt.orNull
  }

  // Test 12: Try.get without recovery
  def parseNumber(s: String): Int = {
    import scala.util.Try
    // VULNERABLE: get on Try can throw
    Try(s.toInt).get
  }

  // Test 13: Either.right.get
  def getValue(either: Either[String, Int]): Int = {
    // VULNERABLE: get on Right projection can throw
    either.toOption.get
  }

  // Test 14: regex match without check
  def extractNumber(input: String): String = {
    val regex = "(\\d+)".r
    input match {
      // VULNERABLE: Pattern might not match
      case regex(num) => num
    }
  }

  // Test 15: Iterator.next without hasNext
  def getNext(items: Iterator[String]): String = {
    // VULNERABLE: No hasNext check
    items.next()
  }

  private def findUser(userId: Int): Option[User] = None
  private def loadConfig(): Map[String, String] = Map.empty
}

case class User(name: String)

trait JavaObject {
  def getName: String
}
